plutus-scb/src/Plutus/SCB/Monitoring.hs

```diff
@@ -0,0 +1,138 @@
+import           Control.Monad.Freer
+import           Control.Monad.Freer.Log        (LogMsg (..), LogObserve (..), Observation (..))
+import qualified Control.Monad.Freer.Log        as L
+import           Control.Monad.IO.Class         (MonadIO (..))
+import           Data.Bifunctor                 (Bifunctor (..))
+import           Data.Functor.Contravariant     (Contravariant (..))
+import           Data.Maybe                     (fromMaybe)
+import           Data.Text                      (Text)
+
+-- | A default 'CM.Configuration' that logs on 'Info' and above
+--   to stdout
+defaultConfig :: IO CM.Configuration
+defaultConfig = do
+  c <- CM.empty
+  CM.setMinSeverity c Info
+  CM.setSetupBackends c [ KatipBK
+                        , AggregationBK
+                        , MonitoringBK
+                        , EKGViewBK
+                        ]
+  CM.setDefaultBackends c [KatipBK, AggregationBK, EKGViewBK]
+  CM.setSetupScribes c [ ScribeDefinition {
+                          scName = "stdout"
+                        , scKind = StdoutSK
+                        , scFormat = ScText
+                        , scPrivacy = ScPublic
+                        , scRotation = Nothing
+                        }]
+  let observables = (Just $ ObservableTraceSelf [MonotonicClock, MemoryStats])
+  CM.setSubTrace c "processAllContractOutboxes" observables
+  CM.setDefaultScribes c ["StdoutSK::stdout"]
+  CM.setEKGport c 12790
+  pure c
+
+-- | Load a 'CM.Configuration' from a YAML file.
+loadConfig :: FilePath -> IO CM.Configuration
+loadConfig = setup
+
+-- | Handle the 'LogMsg' effect by logging messages to a 'Trace'
+handleLogMsgTrace :: forall a m effs.
+  ( LastMember m effs
+  , MonadIO m
+  )
+  => Trace m a
+  -> Eff (LogMsg a ': effs)
+  ~> Eff effs
+handleLogMsgTrace trace = interpret $ \case
+  LMessage L.LogMessage{L._logLevel, L._logMessageContent} ->
+    let defaultPrivacy = Public -- TODO: Configurable / add to 'L.LogMessage'?
+    in sendM $ traceNamedItem trace defaultPrivacy (toSeverity _logLevel) _logMessageContent
+
+handleLogMsgTraceMap :: forall b a m effs.
+  ( LastMember m effs
+  , MonadIO m
+  )
+  => (b -> a)
+  -> Trace m a
+  -> Eff (LogMsg b ': effs)
+  ~> Eff effs
+handleLogMsgTraceMap f t = handleLogMsgTrace (contramap (second (fmap f)) t)
+
+toSeverity :: L.LogLevel -> Severity
+toSeverity = \case
+  L.Debug -> Debug
+  L.Info  -> Info
+  L.Notice -> Notice
+  L.Warning -> Warning
+  L.Error -> Error
+  L.Critical -> Critical
+  L.Alert -> Alert
+  L.Emergency -> Emergency
+
+-- | Handle the 'LogObserve' effect using the 'Cardano.BM.Observer.Monadic'
+--   observer functions
+handleObserveTrace ::
+  forall effs m a.
+  ( LastMember m effs
+  , MonadIO m
+  , MonadCatch m
+  )
+  => CM.Configuration
+  -> Trace m a
+  -> Eff (LogObserve (L.LogMessage Text) ': effs)
+  ~> Eff effs
+handleObserveTrace config t =
+
+  let observeBefore :: (L.LogMessage Text) -> Eff effs (Maybe (SubTrace, CounterState))
```

Comment:
This all seems a bit tricky, could do with some comments.

Reply:
done.